central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -3,18 +3,120 @@ package datastore
 import (
 	"context"
 
+	"github.com/heimdalr/dag"
 	"github.com/stackrox/rox/central/resourcecollection/datastore/index"
 	"github.com/stackrox/rox/central/resourcecollection/datastore/search"
-	"github.com/stackrox/rox/central/resourcecollection/datastore/store/postgres"
+	"github.com/stackrox/rox/central/resourcecollection/datastore/store"
 	v1 "github.com/stackrox/rox/generated/api/v1"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/logging"
 	pkgSearch "github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var (
+	log           = logging.LoggerForModule()
+	initBatchSize = 20
 )
 
 type datastoreImpl struct {
-	storage  postgres.Store
+	storage  store.Store
 	indexer  index.Indexer
 	searcher search.Searcher
+
+	graphLock sync.Mutex
+	graph     *dag.DAG
+}
+
+func resetLocalGraph(ds *datastoreImpl) {
+
+	ds.graphLock.Lock()
+	defer ds.graphLock.Unlock()
+
+	if ds.graph != nil {
+		ds.graph = nil
+	}
+}
+
+func (ds *datastoreImpl) initGraph(ctx context.Context) error {
+	if ds.graph != nil {
+		return nil
+	}
+
+	ds.graphLock.Lock()
+	defer ds.graphLock.Unlock()
+
+	ds.graph = dag.NewDAG()
+
+	// add ids first
+	ids, err := ds.storage.GetIDs(ctx)
+	if err != nil {
+		return err
+	}
+	for _, id := range ids {
+		err = ds.graph.AddVertexByID(id, id)
+		if err != nil {
+			return err
+		}
+	}
+
+	// then add edges by batches
+	for i := 0; i < len(ids); i += initBatchSize {
+		var objs []*storage.ResourceCollection
+		if i+initBatchSize < len(ids) {
+			objs, _, err = ds.storage.GetMany(ctx, ids[i:i+initBatchSize])
+		} else {
+			objs, _, err = ds.storage.GetMany(ctx, ids[i:])
+		}
+		if err != nil {
+			return err
+		}
+		for _, obj := range objs {
+			for _, edge := range obj.GetEmbeddedCollections() {
+				err = ds.graph.AddEdge(obj.GetId(), edge.GetId())
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (ds *datastoreImpl) addCollectionToGraph(ctx context.Context, obj *storage.ResourceCollection) error {
+	if err := ds.initGraph(ctx); err != nil {
+		return err
+	}
+
+	ds.graphLock.Lock()
+	defer ds.graphLock.Unlock()
+
+	err := ds.graph.AddVertexByID(obj.GetId(), obj.GetId())
+	if err != nil {
+		return err
+	}
+	for _, edge := range obj.GetEmbeddedCollections() {
+		err = ds.graph.AddEdge(obj.GetId(), edge.GetId())
+		if err != nil {
+			deleteErr := ds.graph.DeleteVertex(obj.GetId())
+			if deleteErr != nil {
+				log.Errorf("Failed to delete collection, might result in bad state (%v)", deleteErr)
```

Comment:
[nit] "Failed to remove collection from internal state object" (to keep consistent with other messages)?